src/coprocessor/codec/batch/buffer_vec.rs

```diff
@@ -0,0 +1,785 @@
+            // 2
+            for offset in &mut self.offsets {
+                *offset -= data_offset;
+            }
+            assert_eq!(self.offsets[0], 0);
+        } else {
+            self.clear();
+        }
+    }
+
+    /// Shortens the `BufferVec`, keeping the first `n` buffers and dropping the rest.
+    ///
+    /// If `n` >= current length, this has no effect.
+    #[inline]
+    pub fn truncate(&mut self, n: usize) {
+        let len = self.len();
+        if n < len {
+            self.data.truncate(self.offsets[n]);
+            self.offsets.truncate(n);
+        }
+    }
+
+    /// Removes all buffers.
+    #[inline]
+    pub fn clear(&mut self) {
+        self.data.clear();
+        self.offsets.clear();
+    }
+
+    /// Copies all buffers from `other` to `self`.
+    pub fn extend(&mut self, other: &BufferVec) {
+        let base_offset = self.data.len();
+        self.data.extend_from_slice(&other.data);
+        self.offsets.reserve(other.offsets.len());
+        for offset in &other.offsets {
+            self.offsets.push(*offset + base_offset);
+        }
+    }
+
+    /// Copies first `n` buffers from `other` to `self`.
+    ///
+    /// If `n > other.len()`, only `other.len()` buffers will be copied.
+    pub fn extend_n(&mut self, other: &BufferVec, n: usize) {
+        if n >= other.len() {
+            self.extend(other);
+        } else if n > 0 {
+            let base_offset = self.data.len();
+            self.data.extend_from_slice(&other.data[..other.offsets[n]]);
+            self.offsets.reserve(n);
+            for offset in &other.offsets[..n] {
+                self.offsets.push(*offset + base_offset);
+            }
+        }
+    }
+
+    /// Retain the elements according to a boolean array.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `retain_arr` is not long enough.
+    pub fn retain_by_array(&mut self, retain_arr: &[bool]) {
+        // TODO: Optimize to move multiple contiguous slots.
+        unsafe {
+            let len = self.len();
+            assert!(retain_arr.len() >= len);
+
+            if len > 0 {
+                let mut data_write_offset = 0;
+                let mut offsets_write_offset = 0;
+                for i in 0..len - 1 {
+                    if retain_arr[i] {
+                        let write_len = self.offsets[i + 1] - self.offsets[i];
+                        std::ptr::copy(
+                            self.data.as_ptr().add(self.offsets[i]),
+                            self.data.as_mut_ptr().add(data_write_offset),
+                            write_len,
+                        );
+                        self.offsets[offsets_write_offset] = data_write_offset;
+                        offsets_write_offset += 1;
+                        data_write_offset += write_len;
+                    }
+                }
+                // The last buffer does not have an ending offset
+                if retain_arr[len - 1] {
```

Comment:
L220-L232 can merge to L208 with an if expression, e.g: `let write_len = if i == len-1 {self.data.len() - self.offsets[i]} else {self.offsets[i + 1] - self.offsets[i]};`

Reply:
This if expression will be checked every time even for `0 .. len - 1`, which should be less effective than current impl.

Reply:
This if expression will be checked every time even for `0 .. len - 1`, which should be less effective than current impl.

Reply:
OK